core/crypto/rng_hw.c

```diff
@@ -3,19 +3,89 @@
 
 #include <compiler.h>
 #include <crypto/crypto.h>
+#include <kernel/mutex.h>
 #include <kernel/panic.h>
+#include <malloc.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 #include <types_ext.h>
 
-/* This is a HW RNG, no need for seeding */
+static uint8_t *rng_fifo;
+static size_t rng_fifo_size;
+static size_t rng_fifo_pos;
+static struct mutex rng_fifo_mutex = MUTEX_INITIALIZER;
+
+TEE_Result __weak hw_get_max_available_entropy(size_t *blen)
+{
+	*blen = sizeof(size_t);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak hw_get_available_entropy(void *buf)
+{
+	TEE_Result ret;
+
+	ret = hw_get_random_bytes(buf, sizeof(size_t));
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak hw_get_random_bytes(void *buf, size_t blen)
+{
+	uint8_t *buffer = (uint8_t *)buf;
+	size_t buffer_pos = 0;
+	TEE_Result ret = TEE_SUCCESS;
+
+	mutex_lock(&rng_fifo_mutex);
+
+	while (buffer_pos < blen) {
+		/* Refill our FIFO */
+		if (rng_fifo_pos == 0) {
+			while (true) {
+				ret = hw_get_available_entropy(rng_fifo);
+				if (ret == TEE_SUCCESS)
+					break;
+				else if (ret != TEE_ERROR_BUSY)
+					goto out;
+			}
+		}
+
+		buffer[buffer_pos++] = rng_fifo[rng_fifo_pos++];
+		if (rng_fifo_pos == rng_fifo_size)
+			rng_fifo_pos = 0;
+	}
+
+out:
+	mutex_unlock(&rng_fifo_mutex);
+	return ret;
+}
+
 TEE_Result crypto_rng_init(const void *data __unused,
 			   size_t dlen __unused)
 {
+	TEE_Result ret;
+
+	/*
+	 * We do not need to seed our HW RNG, but we do need to allocate
+	 * a buffer for storing extra entropy, set that up here.
+	 */
+	ret = hw_get_max_available_entropy(&rng_fifo_size);
+	if (ret != TEE_SUCCESS) {
+		rng_fifo_size = 0;
+		return ret;
+	}
+
+	rng_fifo = malloc(rng_fifo_size);
```

Comment:
I don't think this is needed. No need to cache the available entropy.

Reply:
We need to store the extra entropy somewhere. That is the core of the performance boost. The HW produces a different amount per cycle than the SW consumes. Without this we have to throw away the extra entropy every time. All the drivers already do this, all we are doing here is centralizing it so the logic is not repeated in each driver.

Reply:
Some drivers have bigger FIFO and we should not wait the HW FIFO is full to provide random byte.

Reply:
With your suggested `hw_get_available_entropy()` we do not wait.

But this FIFO here is used by the function `crypto_rng_read()`/`hw_get_random_bytes()` waiting until the requested amount of random bytes is ready is part of the API.